staging/src/k8s.io/client-go/features/kube_features.go

```diff
@@ -0,0 +1,81 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package features
+
+import (
+	"sync/atomic"
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/component-base/featuregate"
+)
+
+// DefaultFeatureGates returns the feature gates exposed by this library.
+//
+// By default, only the default features gate will be returned.
+// The default implementation allows controlling the features
+// via environmental variables.
+// For example, if you have a feature named "MyFeature,"
+// setting an environmental variable "KUBE_FEATURE_MyFeature"
+// will allow you to configure the state of that feature.
+//
+// Please note that the actual set of the features gates
+// might be overwritten by calling SetFeatureGates method.
+func DefaultFeatureGates() Reader {
```

Comment:
@benluddy is going to see what it would take to use it more like accessing a variable, i.e. `DefaultFeatureGates.Enabled(key)`

Reply:
> @benluddy is going to see what it would take to use it more like accessing a variable, i.e. `DefaultFeatureGates.Enabled(key)`

What is the benefit of doing that?

Reply:
That's not exactly what I was thinking, the specific concern is:

- the singleton can be replaced
- anyone can get a reference to the singleton at a point in time
- the only reason to get a reference to the singleton is to check whether or not a feature is enabled

So if a caller holds onto the object returned from this function before someone else calls `SetFeatureGates`, they might see a different effective value for the same feature than other callers see.

This kind of misuse could be made impossible by design if the package maintains one proxy Reader, changing this function to only return references to that unique object, and changing SetFeatureGates to atomically mutate the Reader behind the proxy. It could be also made impossible with a package-scoped Enabled function, but I think that would be less convenient for testing.